kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtComponentServiceImpl.java

```diff
@@ -244,12 +250,20 @@ public GwtConfigComponent findWireComponentConfigurationFromPid(GwtXSRFToken xsr
                 if (conf != null) {
                     conf.getConfigurationProperties().put(ConfigurationAdmin.SERVICE_FACTORYPID, factoryPid);
                 }
-                if ((conf != null) && (conf.getDefinition() == null)) {
+                if (conf != null && conf.getDefinition() == null) {
                     String temporaryName = String.valueOf(System.nanoTime());
                     cs.createFactoryConfiguration(factoryPid, temporaryName, extraProps, false);
                     try {
-                        // wait for the services to be up
-                        TimeUnit.MILLISECONDS.sleep(500);
+
```

Comment:
Yes, you are right. `Thread.sleep` will not guarantee that the services will be up and running, nor the `waitForService(long)`. As far as the features requires, the best fit would be to go with `CountDownLatch` or `Cyclic Barrier` that would definitely guarantee that if and only if the service is up and running, the Worker Thread that is working in the asynchronous call in GWT, will halt and it will not continue if the temporary factory service is not up and running in the runtime.